Uplift the `let_underscore` lints from clippy into rustc.

a2aaron opened the pull request:
This PR resolves #97241. 

This PR adds three lints from clippy--`let_underscore_drop`, `let_underscore_lock`, and `let_underscore_must_use`, which are meant to capture likely-incorrect uses of `let _ = ...` bindings (in particular, doing this on a type with a non-trivial `Drop` causes the `Drop` to occur immediately, instead of at the end of the scope. For a type like `MutexGuard`, this effectively releases the lock immediately, which is almost certainly the wrong behavior)

In porting the lints from clippy I had to copy over a bunch of utility functions from `clippy_util` that these lints also relied upon. Is that the right approach?

Note that I've set the `must_use` and `drop` lints to Allow by default and set `lock` to Deny by default (this matches the same settings that clippy has). In talking with @estebank he informed me to do a Crater run (I am not sure what type of Crater run to request here--I think it's just "check only"?)

On the linked issue, there's some discussion about using `must_use` and `Drop` together as a heuristic for when to warn--I did not implement this yet.

r? @estebank

a2aaron committed:
Add `let_underscore_drop` lint.

This lint checks for statements similar to `let _ = foo`, where `foo` is
a type that implements `Drop`. These types of let statements cause the
expression in them to be dropped immediately, instead of at the end of
the scope. Such behavior can be surprizing, especially if you are
relying on the value to be dropped at the end of the scope. Instead, the
binding should be an underscore prefixed name (like `_unused`) or the
value should explicitly be passed to `std::mem::drop()` if the value
really should be dropped immediately.

a2aaron committed:
Add `let_underscore_lock` lint.

Similar to `let_underscore_drop`, this lint checks for statements similar
to `let _ = foo`, where `foo` is a lock guard. These types of let
statements are especially problematic because the lock gets released
immediately, instead of at the end of the scope. This behavior is almost
always the wrong thing.

a2aaron committed:
Add `let_underscore_must_use` lint.

Similar to `let_underscore_drop`, this lint checks for statements similar
to `let _ = foo`, where `foo` is an expression marked `must_use`.

a2aaron committed:
Move let_underscore tests to their own subfolder.

This was done to pass `tidy`.

a2aaron committed:
Allow `let_underscore_drop` and `let_underscore_must_use` by default.

These lints are very noisy and are allow-by-default in clippy anyways.
Hence, setting them to allow-by-default here makes more sense than
warning constantly on these cases.

a2aaron committed:
Show code suggestions in `let_undescore` lint messages.

This commit uses `span_suggestion_verbose` to add what specific code
changes can be done as suggested by the lint--in this case, either binding
the expression to an unused variable or using `std::mem::drop` to drop
the value explicitly.

a2aaron committed:
Set `let_underscore_lock` to Deny by default.

Clippy sets this lint to Deny by default, and it having the lint be Deny
is useful for when we test the lint against a Crater run.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @estebank (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6740107612?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/2db2813ed8dee75acf3e5dc8e8a07dc37776635d/checks/6740107612/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
running 59 tests
.........ii....................F...........................
failures:

---- src/let_underscore.rs - let_underscore::LET_UNDERSCORE_LOCK (line 57) stdout ----
error[E0433]: failed to resolve: use of undeclared crate or module `thread`
 --> src/let_underscore.rs:61:1
  |
6 | thread::spawn(move || {
  | ^^^^^^ use of undeclared crate or module `thread`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0433`.
Couldn't compile the test.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6740332273?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/aeb8ca20dcfa2df70542c74a8cce7d8dc113d43c/checks/6740332273/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
running 59 tests
.........ii.......F........................................
failures:

---- src/let_underscore.rs - let_underscore::LET_UNDERSCORE_LOCK (line 57) stdout ----
error: non-binding let on a synchronization lock
  --> src/let_underscore.rs:65:5
   |
10 |     let _ = data.lock().unwrap();
   |
   |
   = note: `#[deny(let_underscore_lock)]` on by default
help: consider binding to an unused variable
   |
10 |     let _unused = data.lock().unwrap();
help: consider explicitly droping with `std::mem::drop`
   |
   |
10 |     let _ = drop(...);

error: aborting due to previous error

Couldn't compile the test.
```

</details>

cjgillot commented on compiler/rustc_lint/src/let_underscore.rs:
```diff
@@ -0,0 +1,270 @@
+        fn has_must_use_attr(cx: &LateContext<'_>, did: hir::def_id::DefId) -> bool {
+            cx.tcx
+                .get_attrs(did, rustc_span::sym::must_use)
+                .find(|a| a.has_name(rustc_span::sym::must_use))
```
Isn't this `has_name` call trivially true?
IIUC, this construct only tests that the return of `get_attrs` is not empty.
Can `cx.tcx.has_attr(did, rustc_span::sym::must_use)` be used?

cjgillot commented on src/test/ui/let_underscore/let_underscore_drop.rs:
```diff
@@ -0,0 +1,14 @@
+// run-pass
```
Can you move the `let_underscore` directory inside `src/test/ui/lint`?

cjgillot commented on compiler/rustc_lint/src/let_underscore.rs:
```diff
@@ -0,0 +1,270 @@
+const SYNC_GUARD_PATHS: [&[&str]; 5] = [
+    &["std", "sync", "mutex", "MutexGuard"],
+    &["std", "sync", "rwlock", "RwLockReadGuard"],
+    &["std", "sync", "rwlock", "RwLockWriteGuard"],
```
Can those paths be replaced by diagnostic items?  This will make all this more robust to code being moved within `std`.

a2aaron replied:
These paths were taken directly from the clippy implementation (which is also why they're hardcoded right now--there is a [clippy issue](https://github.com/rust-lang/rust-clippy/issues/8859) meant to address that for clippy though). I can add a diagnostic item for these types, but I'm wondering if something similar to clippy's `has_significant_drop` already exists? It would make the lint much easier to implement and would generalize to other types well.

For now, however, I've added some diagnostic items.

cjgillot commented on compiler/rustc_lint/src/let_underscore.rs:
```diff
@@ -0,0 +1,270 @@
+    &["std", "sync", "rwlock", "RwLockReadGuard"],
+    &["std", "sync", "rwlock", "RwLockWriteGuard"],
+    &["parking_lot", "raw_mutex", "RawMutex"],
+    &["parking_lot", "raw_rwlock", "RawRwLock"],
```
I don't know the policy here: is `rustc` supposed to warn on inappropriate use of non-std crates?

a2aaron replied:
I'll remove the `parking_lot` specific paths for now, unless someone wants to keep them in.

cjgillot commented on compiler/rustc_lint/src/let_underscore.rs:
```diff
@@ -0,0 +1,270 @@
+        }
+        if let Some(init) = local.init {
+            let init_ty = cx.typeck_results().expr_ty(init);
+            let needs_drop = init_ty.needs_drop(cx.tcx, cx.param_env);
```
Can we exit early if `needs_drop` is false?  The later computations seem a bit heavy.

a2aaron replied:
Bailing out early would mean that types which are `must_use` but dont impl `Drop` won't get checked. Doing this would probably make `let_underscore_must_use` not very useful. (I am okay with just only having `let_underscore_lock` as lint, given that Clippy sets `let_underscore_must_use` and `let_underscore_drop` as allow-by-default)

a2aaron replied:
I decided to just remove `let_underscore_must_use` and implemented the early bailout instead.

cjgillot commented on compiler/rustc_lint/src/let_underscore.rs:
```diff
@@ -0,0 +1,270 @@
+                .span_suggestion_verbose(
+                    init_span,
+                    "consider explicitly droping with `std::mem::drop`",
+                    "drop(...)",
```
Should we craft an applicable suggestion here, or do we expect low usefulness?

a2aaron replied:
I don't know how to do this--I tried using `format!("drop({})", init)`  to create an applicable suggestion, but I couldn't figure out how to turn the `init` expression into text.

a2aaron replied:
Update: I was able to figure out how to create a suggestion using `multipart_suggestion`. I wasn't sure how to remove the `let _ =` part of the message however (In the docs, I can't seem to figure out where the `let` and `=` parts of a `Local` statement are kept in the span)

cjgillot reviewed (commented):
Thanks @a2aaron. I left a few questions.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6740659551?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/6b179e3a67a2242b63892a1831ba217d968c2f62/checks/6740659551/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling tempfile v3.2.0
   Compiling serde_json v1.0.59
   Compiling lint-docs v0.1.0 (/checkout/src/tools/lint-docs)
    Finished release [optimized] target(s) in 5.79s
warning: lint group `let-underscore` is missing from the GROUP_DESCRIPTIONS list
If this is a new lint group, please update the GROUP_DESCRIPTIONS in src/tools/lint-docs/src/groups.rs
Rustbook (x86_64-unknown-linux-gnu) - cargo
Rustbook (x86_64-unknown-linux-gnu) - embedded-book
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
---
doc tests for: /checkout/src/doc/rustc/src/tests/index.md
doc tests for: /checkout/src/doc/rustc/src/what-is-rustc.md
 finished in 1.184 seconds
Generating lint docs (x86_64-unknown-linux-gnu)
error: failed to test example in lint docs for `let_underscore_drop` in /checkout/compiler/rustc_lint/src/let_underscore.rs:10: lint docs should contain the line `{{produces}}`
This error was generated by the lint-docs tool.
This error was generated by the lint-docs tool.
This tool extracts documentation for lints from the source code and places
them in the rustc book. See the declare_lint! documentation
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/macro.declare_lint.html

Build completed unsuccessfully in 0:32:58
Build completed unsuccessfully in 0:32:58
To re-run these tests, run: ./x.py test --keep-stage=0 src/tools/lint-docs
The --keep-stage flag should be used if you have already built the compiler

```

</details>

a2aaron committed:
Move local functions to outer scope.

a2aaron committed:
Move `let_underscore` tests into the `lint` subfolder.

a2aaron committed:
Add `let_underscore` lint group to `GROUP_DESCRIPTIONS`.

a2aaron committed:
Use `has_attr` instead of `get_attrs` in `has_must_use_attr`

a2aaron committed:
Bail out early if the type does not has a trivial Drop implementation.

If the type has a trivial Drop implementation, then it is probably irrelevant
that the type was dropped immediately, since nothing important
happens on drop. Hence, we can bail out early instead of doing some
expensive checks.

a2aaron committed:
Use diagnostic items instead of hard coded paths for `let_underscore_lock`

Using diagnostic items avoids having to update the paths if the guard
types ever get moved around for some reason. Additionally, it also greatly
simplifies the `is_sync_lock` check.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6742186342?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/6342b58ef0f71ca0284dced4b1d22f9726c1c74a/checks/6742186342/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
........................................................................................ 13288/13303
...............
failures:

---- [ui] src/test/ui/lint/let_underscore/let_underscore_lock.rs stdout ----
error: ui test compiled successfully!
status: exit status: 0
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/lint/let_underscore/let_underscore_lock.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/let_underscore/let_underscore_lock" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/let_underscore/let_underscore_lock/auxiliary"
stdout: none
stderr: none


---- [ui] src/test/ui/lint/let_underscore/let_underscore_must_use.rs stdout ----


- warning: non-binding let on a expression marked `must_use`
-   --> $DIR/let_underscore_must_use.rs:11:5
-    |
- LL |     let _ = MustUseType;
-    |
-    |
-    = note: requested on the command line with `-W let-underscore-must-use`
- help: consider binding to an unused variable
-    |
- LL |     let _unused = MustUseType;
- help: consider explicitly droping with `std::mem::drop`
-    |
-    |
- LL |     let _ = drop(...);
- 
- 
- warning: non-binding let on a expression marked `must_use`
-   --> $DIR/let_underscore_must_use.rs:12:5
- LL |     let _ = must_use_function();
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-    |
- help: consider binding to an unused variable
- help: consider binding to an unused variable
-    |
- LL |     let _unused = must_use_function();
- help: consider explicitly droping with `std::mem::drop`
-    |
-    |
- LL |     let _ = drop(...);
- 
- warning: 2 warnings emitted
- 
- 
---
To only update this specific test, also pass `--test-args lint/let_underscore/let_underscore_must_use.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/lint/let_underscore/let_underscore_must_use.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/let_underscore/let_underscore_must_use/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-W" "let_underscore_must_use" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/let_underscore/let_underscore_must_use/auxiliary"
stdout: none
stderr: none


failures:
    [ui] src/test/ui/lint/let_underscore/let_underscore_lock.rs
```

</details>

joshtriplett commented:
:+1: for adding `let_underscore_lock` and making it `deny` by default; that seems likely to catch bugs.

Adding `let_underscore_drop` and making it `allow` by default seems fine as well; I don't think we ever want to make that `warn` or `error`, since it's legitimate code, but I can imagine someone wanting to find and catch such code.

However, I'm not sure I understand the value of `let_underscore_must_use`. `let _ = ...` seems like exactly the pattern people normally use to *intentionally* ignore a `must_use`; under what circumstances would someone want to get a warning or error about this?

a2aaron commented:
> However, I'm not sure I understand the value of `let_underscore_must_use`. `let _ = ...` seems like exactly the pattern people normally use to _intentionally_ ignore a `must_use`; under what circumstances would someone want to get a warning or error about this?

I added it to match the corresponding clippy lint, but I think you're right--if people really want it, clippy would probably be a more appropriate tool for that, I'll remove `let_underscore_must_use`.

a2aaron committed:
Use `check-pass` instead of `run-pass`

We don't actually care about running these programs, only checking the
warnings they generate.

a2aaron committed:
Remove `let_underscore_must_use`

The `let_underscore_must_use` lint was really only added because clippy
included it, but it doesn't actually seem very useful.

a2aaron committed:
Add diagnostic items to MutexGuard and RwLock Guards

I forgot to add the diagnostic to the actual types in `std` earlier.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6742891783?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/321a598b7519733aff7dea8b4ebd3631626d656b/checks/6742891783/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
doc tests for: /checkout/src/doc/rustc/src/tests/index.md
doc tests for: /checkout/src/doc/rustc/src/what-is-rustc.md
 finished in 1.181 seconds
Generating lint docs (x86_64-unknown-linux-gnu)
error: failed to test example in lint docs for `let_underscore_drop` in /checkout/compiler/rustc_lint/src/let_underscore.rs:7: lint docs should contain the line `{{produces}}`
This error was generated by the lint-docs tool.
This error was generated by the lint-docs tool.
This tool extracts documentation for lints from the source code and places
them in the rustc book. See the declare_lint! documentation
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/macro.declare_lint.html

Build completed unsuccessfully in 0:32:35
Build completed unsuccessfully in 0:32:35
To re-run these tests, run: ./x.py test --keep-stage=0 src/tools/lint-docs
The --keep-stage flag should be used if you have already built the compiler

```

</details>

a2aaron committed:
Add `{{produces}}` tag to lint doc comments.

a2aaron committed:
Use `multipart_suggestion` to create an applicable suggestion.

The "consider explicitly droping" can now suggest a machine applicable
suggestion now.

estebank commented:
@bors try @craterbot check

bors commented:
:hourglass: Trying commit cdf66060666fa53cbca9647ac7434bdfd2cc37a5 with merge 844341966fa4deedcf5964b5cbcd30dd60c394b4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"cdf66060666fa53cbca9647ac7434bdfd2cc37a5","merge_sha":"844341966fa4deedcf5964b5cbcd30dd60c394b4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6823863743?check_suite_focus=true)
Build commit: 844341966fa4deedcf5964b5cbcd30dd60c394b4 (`844341966fa4deedcf5964b5cbcd30dd60c394b4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6823863743?check_suite_focus=true"},"merge_sha":"844341966fa4deedcf5964b5cbcd30dd60c394b4"} -->

estebank commented on compiler/rustc_lint/src/let_underscore.rs:
```diff
@@ -0,0 +1,166 @@
+    /// calling `std::mem::drop` on the expression is clearer and helps convey
+    /// intent.
+    pub LET_UNDERSCORE_DROP,
+    Allow,
```
Ah! Can you push a new commit with this being `Deny`? That way we can do a crater run to see how bad this heuristic would actually be in the wild. Depending on the results of that run, we'd make this either `Warn` or `Allow`, as you have here.

@joshtriplett this is necessary to catch *all* cases of the RAII pattern, but we need to verify what the S/N ratio is for the simplistic heuristic. Otherwise we'd need to extend `LET_UNDERSCORE_LOCK` to also check for the mutexes in parkinglot and tokio, at least.

a2aaron replied:
Done! I will note that this ends up triggering on code in the compiler itself, which I'm guessing means it's going to be extremely noisy.

joshtriplett replied:
@estebank I would hypothesize that `let _ = returns_type_implementing_drop()` is quite common, too much so to turn on by default.

estebank replied:
Case in point:

```
   Compiling rustc-demangle v0.1.21
error: non-binding let on a type that implements `Drop`
   --> library/alloc/src/vec/into_iter.rs:324:21
    |
324 |                     let _ = RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc);
    |
    |
    = note: `#[deny(let_underscore_drop)]` on by default
help: consider binding to an unused variable to avoid immediately dropping the value
    |
324 |                     let _unused = RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc);
help: consider immediately dropping the value
    |
    |
324 |                     drop(RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc));

error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:04:32
```

:-/

estebank replied:
@a2aaron this is a problem for what I wanted to do. Did you see how many cases are in the rustc repo? Because if `rustc` doesn't build, then we can't run crater to collect numbers on how many crates would suddenly start warning on this. I guess path inspecting against a denylist of common crates (or better, stabilizing an attribute that they can annotate with that `rustc` checks for) might be warranted in the end :-/

a2aaron replied:
@estebank There's a lot. Setting `let_underscore_drop` to `Warn` and building with `--warnings warn` turned on, I got 28 warnings. Most of these were in `std`. It seems like 18 of them are caused by one macro (`rtprintpanic!`).

estebank replied:
That's actually not as many as I thought there would be! 😅

a2aaron replied:
Should I try manually fixing up all of those cases?

estebank replied:
Only if you have time, otherwise we can defer on that and focus on the specific denylist of std mutexes.

a2aaron replied:
I think I'll defer on it--the lint is too noisy and annoying to fix up right now, and every instance of it firing has been on types where it really doesn't matter that it gets dropped earlier (it's firing on a lot of `Box`s or various `Result` types). I've pushed a commit making `let_underscore_drop` allow by default for now.

estebank commented on src/test/ui/lint/let_underscore/let_underscore_drop.stderr:
```diff
@@ -0,0 +1,18 @@
+   |
+LL |     let _unused = NontrivialDrop;
+   |         ~~~~~~~
+help: consider explicitly droping with `std::mem::drop`
+   |
+LL |     let _ = drop(NontrivialDrop);
+   |             +++++              +
```
As mentioned already, this should be

```suggestion
help: consider explicitly droping with `std::mem::drop`
   |
LL |     drop(NontrivialDrop);
   |     ~~~~~              +
```

a2aaron committed:
Reword suggestion messages.

a2aaron committed:
Have the drop code suggestion not include `let _ =`

a2aaron committed:
Make `let_underscore_drop` Deny by default.

This is done so that we can check the noisiness of this lint in a Crater
run. Note that when I built the compiler, I actually encountered lots of
places where this lint will trigger and fail compilation, so I had to
also set `RUSTFLAGS_NOT_BOOSTRAP` to `-A let_underscore_drop` when
compiling to prevent that.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6844013016?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/8807c2d8e512668a6f09ddb14c6742bf28e66816/checks/6844013016/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc-demangle v0.1.21
error: non-binding let on a type that implements `Drop`
   --> library/alloc/src/vec/into_iter.rs:324:21
    |
324 |                     let _ = RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc);
    |
    |
    = note: `#[deny(let_underscore_drop)]` on by default
help: consider binding to an unused variable to avoid immediately dropping the value
    |
324 |                     let _unused = RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc);
help: consider immediately dropping the value
    |
    |
324 |                     drop(RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc));

error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:04:32
```

</details>

a2aaron committed:
Re-allow `let_underscore_drop` by default.

This lint is way way too noisy to have it be `Deny` by default.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6928866902?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5b2b74850de9038872570f08fa661804c886e352/checks/6928866902/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc-demangle v0.1.21
error: non-binding let on a type that implements `Drop`
   --> library/alloc/src/vec/into_iter.rs:324:21
    |
324 |                     let _ = RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc);
    |
    |
    = note: `-D let-underscore-drop` implied by `-D warnings`
help: consider binding to an unused variable to avoid immediately dropping the value
    |
324 |                     let _unused = RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc);
help: consider immediately dropping the value
    |
    |
324 |                     drop(RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc));

error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:03:59
```

</details>

a2aaron commented:
@bors try @craterbot check

bors commented:
@a2aaron: :key: Insufficient privileges: not in try users

a2aaron commented:
@estebank ah it seems i cannot do a crater run. If there aren't any other changes I should make, can you run the crater run?

estebank commented:
@bors try @craterbot check

bors commented:
:hourglass: Trying commit a9095ff2139fb305b15006d1f2a1ff16da0b6c29 with merge 5c7108a2805ef56743d713d2789c5e802631c9e8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a9095ff2139fb305b15006d1f2a1ff16da0b6c29","merge_sha":"5c7108a2805ef56743d713d2789c5e802631c9e8"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/7050127171?check_suite_focus=true)
Build commit: 5c7108a2805ef56743d713d2789c5e802631c9e8 (`5c7108a2805ef56743d713d2789c5e802631c9e8`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/7050127171?check_suite_focus=true"},"merge_sha":"5c7108a2805ef56743d713d2789c5e802631c9e8"} -->

a2aaron commented:
Looks like the crater run finished! The report says 47 crates failed--I checked a few of the failures and they do seem to be accurate (people `let _ =` a mutex).

estebank commented:
@a2aaron looking at the output, it seems indeed to be catching real issues 🎉

Looking at one of them with fresh eyes, I'm realizing that we're not telling people *why* this is a problem:

```
error: non-binding let on a synchronization lock
   --> src/main.rs:106:9
    |
106 |         let _ = lock.lock().unwrap();
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[deny(let_underscore_lock)]` on by default
help: consider binding to an unused variable to avoid immediately dropping the value
    |
106 |         let _unused = lock.lock().unwrap();
    |             ~~~~~~~
help: consider immediately dropping the value
    |
106 |         drop(lock.lock().unwrap());
    |         ~~~~~                    +
```

I think we should add more info to the lint:

```
error: non-binding let on a synchronization lock
   --> src/main.rs:106:9
    |
106 |         let _ = lock.lock().unwrap();
    |             ^   ^^^^^^^^^^^^^^^^^^^^ this lock is not assigned to a binding and is immediately dropped
    |             |
    |             this binding will immediately drop the value assigned to it
    |
    = note: `#[deny(let_underscore_lock)]` on by default
help: consider binding to an unused variable to dropp the value at the end of its scope
    |
106 |         let _unused = lock.lock().unwrap();
    |             ~~~~~~~
help: consider explicitly immediately dropping the value
    |
106 |         drop(lock.lock().unwrap());
    |         ~~~~~                    +
```

In order to get that multiple span primary, you can use `let span: MultiSpan = vec![span1, span2];` and `span.push_span_label(span1, "text".to_string());` for the labels.

a2aaron committed:
Explain why let-underscoring a lock guard is incorrect.

Currently, the let_underscore_lock lint simply tells what is wrong, but
not why it is wrong. We fix this by using a `MultiSpan` to explain
specifically that doing `let _ = ` immediately drops the lock guard
because it does not assign the lock guard to a binding.

a2aaron commented:
@estebank Done! I've added the additional diagnostics to the lint.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7680884089?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/a9f1b7bd2a25e34de29eb88f81550690f4fec5dc/checks/7680884089/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
error[E0603]: struct `LintDiagnosticBuilder` is private
 --> compiler/rustc_lint/src/let_underscore.rs:4:26
  |
4 | use rustc_middle::{lint::LintDiagnosticBuilder, ty};
  |                          ^^^^^^^^^^^^^^^^^^^^^ private struct
note: the struct `LintDiagnosticBuilder` is defined here
 --> /checkout/compiler/rustc_middle/src/lint.rs:5:46
  |
  |
5 | use rustc_errors::{Diagnostic, DiagnosticId, LintDiagnosticBuilder, MultiSpan};

For more information about this error, try `rustc --explain E0603`.
error: could not compile `rustc_lint` due to previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

a2aaron committed:
Fix imports.

I'm not really sure why this is nessecary to do, but the checks on the
PR do not seem to work if do not do this.

estebank commented on src/test/ui/lint/let_underscore/let_underscore_drop.rs:
```diff
@@ -0,0 +1,14 @@
+// check-pass
+// compile-flags: -W let_underscore_drop
```
You can accomplish the same behavior with `#![warn(let_underscore_drop)]`

a2aaron replied:
Done!

estebank commented on compiler/rustc_lint/src/let_underscore.rs:
```diff
@@ -0,0 +1,175 @@
+                return;
+            }
+            let is_sync_lock = match init_ty.kind() {
+                ty::Adt(adt, _) => SYNC_GUARD_SYMBOLS
+                    .iter()
+                    .any(|guard_symbol| cx.tcx.is_diagnostic_item(*guard_symbol, adt.did())),
```
If we were ever to introduce 3rd party checks, it would be an extra arm here.

estebank commented:
@bors r+

bors commented:
:pushpin: Commit d355ec94ff0ef409bfa953f715987a98de13fd64 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank d355ec94ff0ef409bfa953f715987a98de13fd64 -->
<!-- homu: {"type":"Approved","sha":"d355ec94ff0ef409bfa953f715987a98de13fd64","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d355ec94ff0ef409bfa953f715987a98de13fd64 with merge 0fb29cc7184382b8ad840d84e40022dbb15351f2...
<!-- homu: {"type":"BuildStarted","head_sha":"d355ec94ff0ef409bfa953f715987a98de13fd64","merge_sha":"0fb29cc7184382b8ad840d84e40022dbb15351f2"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/7695001669?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/7695001669?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/7694650490?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/0fb29cc7184382b8ad840d84e40022dbb15351f2/checks/7694650490/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
error: RPC failed; curl 56 GnuTLS recv error (-54): Error in the pull function.
fatal: The remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
thread 'main' panicked at 'assertion failed: status.success()', src/tools/cargotest/main.rs:148:13
 finished in 217.073 seconds
Build completed unsuccessfully in 0:18:09
Build completed unsuccessfully in 0:18:09
make: *** [check-aux] Error 1
Makefile:44: recipe for target 'check-aux' failed
```

</details>

a2aaron committed:
Use `#![warn(let_underscore_drop)]` in tests.

estebank commented:
@bors r+

bors commented:
:pushpin: Commit 76c90c3015a7e3ad6f0e6b807839ff59f17eba89 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank 76c90c3015a7e3ad6f0e6b807839ff59f17eba89 -->
<!-- homu: {"type":"Approved","sha":"76c90c3015a7e3ad6f0e6b807839ff59f17eba89","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

detly commented:
The [1.65.0 changelog(https://github.com/rust-lang/rust/blob/stable/RELEASES.md#language) says

> Add lints `let_underscore_drop`, `let_underscore_lock`, and `let_underscore_must_use` from Clippy

But [this comment](https://github.com/rust-lang/rust/pull/97739/#issuecomment-1146734401) and [b5b5b54](https://github.com/rust-lang/rust/pull/97739/commits/b5b5b5471ba324f6b4d5a53957a800ad44126964) seems to indicate that `let_underscore_must_use` was omitted. Is the changelog wrong?
